### PR TITLE
Function arguments should never get promoted

### LINKED ISSUE #59469: Wrapping simd call results in compiler panic (nightly)
Expected the following code to compile

```rust
#[cfg(target_arch = "x86")]
use core::arch::x86::*;
#[cfg(target_arch = "x86_64")]
use core::arch::x86_64::*;

unsafe fn pclmul(a: __m128i, b: __m128i, imm8: i32) -> __m128i {
    _mm_clmulepi64_si128(a, b, imm8)
}
```

But getting

```
Standard Error

   Compiling playground v0.0.1 (/playground)
thread 'rustc' panicked at 'index out of bounds: the len is 2 but the index is 3', /rustc/fbd34efb32b9efb574899e4335bdc8c6525ac27e/src/libcore/slice/mod.rs:2539:10
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.35.0-nightly (fbd34efb3 2019-03-26) running on x86_64-unknown-linux-gnu

note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

Repro: 

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=36e55f5180a9cf903585dead5b4c2a8d

### PR DESCRIPTION
**Pull request by oli-obk**
fixes https://github.com/rust-lang/rust/issues/59469

### CODE REVIEW

File: src/librustc_mir/transform/qualify_consts.rs [hunk 1]
```diff
@@ -639,7 +639,7 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
                             per_local.insert(local);
                         }
                     }
-                    cx.per_local[IsNotPromotable].insert(local);
+                    cx.per_local[IsNotConst].insert(local);
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: I think we should do this for all locals where `if !temps[local].is_promotable()`.

Comment (eddyb) [reply src/librustc_mir/transform/qualify_consts.rs]: What I meant is that we should do this for all locals, not just temps.
That is, don't do it on a case-by-case basis in this `match`, but rather have this after it:
```rust
            if !temps[local].is_promotable() {
                cx.per_local[IsNotConst].insert(local);
            }
```

**Comment by pnkfelix** [comment]
As an experiment, I tried backporting this change to the beta branch, to double-check that it actually addresses the regression.

One potential issue is that the regression tests may not themselves be immediately backportable as written, because I do not think #59044 is in the beta branch. Just a warning.

**Comment by pnkfelix** [comment]
(commit ccc0cd866d9cdd931ccf619b635ea217a4baf29d itself *does* seem to address the key problem in the beta branch.)

@oli-obk @eddyb Regarding the beta-nomination, I do not think we can afford to wait until the next `rustc` meeting. Do you think we should go ahead and beta-accept this now? Or wait until @nikomatsakis is back?

(To be absolutely clear, I have no personal problem with backporting  ccc0cd866d9cdd931ccf619b635ea217a4baf29d to beta.)

**Comment by oli-obk** [comment]
I think we should beta accept this without a full compiler team consensus as it's small and important to get fixed. There seems no alternative to me.

I can create a beta backport that regenerates the ui output the way beta expects it.

File: src/librustc_mir/transform/qualify_consts.rs (hunk 1 above)
Comment (oli-obk) [reply src/librustc_mir/transform/qualify_consts.rs]: That's a change that I don't want to backport. are you alright with r+ing this PR and I'll address this immediately in a master-only PR?

**Comment by Centril** [comment]
> There seems no alternative to me.

Yes; absolutely. I would *really* like to avoid a C-future-compatibility warning due to a failure to backport.

File: src/librustc_mir/transform/qualify_consts.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc_mir/transform/qualify_consts.rs]: There's a risk of there being more issues we haven't hit due to this not being general enough.
This is not just code cleanup, I think right now this is kind of broken.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 2]
```diff
@@ -639,15 +639,15 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
...
                 }
 
                 LocalKind::Var if mode == Mode::Fn => {
                     cx.per_local[IsNotConst].insert(local);
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Like, does this mean we'll try to promote a `Var` in a `const fn`?!

File: src/librustc_mir/transform/qualify_consts.rs [hunk 3]
```diff
@@ -817,15 +817,15 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
...
         }
 
-        // Ensure the `IsNotPromotable` qualification is preserved.
+        // Ensure the `IsNotConst` qualification is preserved.
         // NOTE(eddyb) this is actually unnecessary right now, as
         // we never replace the local's qualif, but we might in
         // the future, and so it serves to catch changes that unset
         // important bits (in which case, asserting `contains` could
         // be replaced with calling `insert` to re-set the bit).
         if kind == LocalKind::Temp {
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Again, here, this seems risky, why don't we do this for all locals?
I think relying on the `Var` / `Temp` distinction was a bit of a mistake... I mean, here, as opposed to in the analysis in `promote_consts`.

**Comment by eddyb** [comment]
r=me with or without https://github.com/rust-lang/rust/pull/59724#discussion_r272788749 / https://github.com/rust-lang/rust/pull/59724#discussion_r272788800 resolved

For runtime code, this PR is probably fine (it marks all args and vars as "not const" - not the return place, but that tends not to ever get borrowed or passed to a function), but for `const fn` some of this pass might still be bogus.

**Comment by pnkfelix** [comment]
> I think we should beta accept this without a full compiler team consensus as it's small and important to get fixed. There seems no alternative to me.

I agree with this and am going to mark the PR as beta-accepted accordingly

**Comment by eddyb** [comment]
I guess I shouldn't delay this needlessly:
@bors r+ p=10 (needs backport soon)

**Comment by kennytm** [comment]
@bors r-

The new test `src/test/ui/consts/const_arg_wrapper.rs` needs to be `// only-x86_64`.

**Comment by pietroalbini** [comment]
@bors retry

Yielding priority to the beta backports.

**Comment by pietroalbini** [comment]
@bors retry

Yielding priority *again* to the beta backports.

**Comment by Centril** [comment]
Moving before the clippy PR, @bors p=15

**Pull request merged**